AF_UNIX support for beanstalkd

etanol opened the pull request:
Add the posibility of accepting connections from a local socket (bound to a filesystem path).

To minimize portability issues, socket paths longer than 100 bytes are rejected.  Systemd support has been added too, but not tested as I don't have access to any Fedora or similar.

The behaviour of the new command line switch may be strange, but it has been implemented in a way to avoid script breakage when this option is not used.

The automatic socket file replacement may be a bit surprising too.  However, I found that implementing an automatic removal/cleanup on exit was a bit more intrusive.

Any feedback will be appreciated.

ShadowOfHarbringer commented:
I would be greatly interested in this feature. 
Some of my servers handle so many connections that i already had problems with too many open ports limit on Linux.

petri commented:
@kr, I also have come to desire this feature. Any chance you could merge it and release? The beanstalkc Python client apparently has already support for unix socket, just waiting for beanstalkd itself to support it, and I would gladly add support to pybeanstalk as well.

silviucpp commented:
Hello @etanol can you please rebase the changes with the last master ?

Silviu

djzort commented:
Any update on this?

etanol commented:
I believe this project is about to transition into *abandoned* status.

sshilko commented:
feels bad, but you are correct, there are many fresh alternatives (NSQ/NATS/....) nsq has very similar API but i liked beanstalkd for stability

etanol commented on doc/beanstalkd.1:
```diff
@@ -57,6 +57,9 @@ Show a brief help message and exit\.
 Listen on address \fIaddr\fR (default is 0\.0\.0\.0)\.
 .
 .IP
+When \fIaddr\fR starts with "unix:", the unprefixed value of will be used as the local filesystem path to create a UNIX socket instead of a TCP socket\. In this case the vale of \fB\-p\fR will be ignored\.
```
How can I fix that from the *ronn* file?  Because this ROFF code was autogenerated.

pprindeville replied:
okay, never mind... the preprocessor is broken.

meox commented:
this is a great PR please rework on it

ysmolsky commented:
@etanol can you please look at it again?

pprindeville commented on net.c:
```diff
@@ -37,104 +215,23 @@ make_server_socket(char *host, char *port)
-
-      break;
+    if (host && strlen(host) > 5 && !strncmp(host, "unix:", 5)) {
+        return make_unix_socket(host + 5);
```
Actually, you could drop the `strlen(host) > 5` altogether... `strncmp()` will do the right thing in all cases.

etanol replied:
You think?  What if the length of the `host` string is less than 5?  Wouldn't `strncmp` read past the boundaries of `host`?

Fudoshiki commented:
Any plan to merge this?

etanol commented:
I will give it one more try.  But this is getting ridiculous, seven years...

Fudoshiki commented:
This is very useful feature

ysmolsky commented:
@etanol thanks for doing this. many things has changed since that time, so thanks again.

ysmolsky commented:
@etanol, thank you. Do you plan to address the review comments by @pprindeville? Do you have a patience for my review comments as well?

etanol commented:
@ysmolsky Sorry for that, I forgot about those amendments.  I have updated the PR with the updates.  I also added some extra tests to keep coverage on check.

pprindeville commented on net.c:
```diff
@@ -7,19 +7,197 @@
+    addr.sun_family = AF_UNIX;
+    if (strlen(path) > maxlen) {
+        warnx("truncating local socket path to: %.*s", (int)maxlen, path);
+        strncpy(addr.sun_path, path, maxlen);
```
What about using strlcpy() instead?

etanol replied:
It's not as standard, and it would require quite a bit of boilerplate to compile it properly under the various UNIX flavours.

ysmolsky replied:
There is no need for strlcpy() here, see below.

pprindeville commented on testserv.c:
```diff
@@ -141,30 +165,39 @@ kill_srvpid(void)
-    srv.sock.fd = make_server_socket("127.0.0.1", "0");
+    if (socket_file != NULL) {
+        char name[200];
+        snprintf(name, 200, "unix:%s", socket_file);
```
sizeof(name) instead of 200...

pprindeville commented on testserv.c:
```diff
@@ -343,6 +376,20 @@ cttest_too_long_commandline()
+    struct sockaddr_un addr;
+    const size_t maxlen = sizeof(addr.sun_path);
+    char real_name[maxlen];
+    snprintf(real_name, maxlen, NAME);
```
There's no %-format specifier, so please don't use snprintf(); use strlcpy() instead.

etanol replied:
Good point, although I'll stick to `snprintf`.  Remember that this is just test code, so the fact that it works triumphs over its elegance.

pprindeville commented on testserv.c:
```diff
@@ -383,6 +430,41 @@ cttest_peek_not_found()
+{
+    progname = __func__;
+    char name[90];
+    snprintf(name, 90, "/tmp/ctb-%d", getpid());
```
sizeof(name) instead of 90...

pprindeville commented on testserv.c:
```diff
@@ -383,6 +430,41 @@ cttest_peek_not_found()
+{
+    progname = __func__;
+    char name[90];
+    snprintf(name, 90, "/tmp/ctb-%d", getpid());
```
sizeof(name)

ysmolsky commented on net.c:
```diff
@@ -7,19 +7,197 @@
+        addr.sun_path[maxlen] = '\0';
+    } else {
+        strcpy(addr.sun_path, path);
+    }
```
There is no need to do copying in the branches. You can leave the warning in the branch and just copy string afterwards. It wont copy more than len of `path` and it will fill the rest of the `addr.sun_path` buffer with zeroes:
```c
    if (strlen(path) > maxlen) {
        warnx("truncating local socket path to: %.*s", (int)maxlen, path);
    }
    strncpy(addr.sun_path, path, maxlen);
```

ysmolsky commented on net.c:
```diff
@@ -7,19 +7,197 @@
+    if (verbose) {
+        printf("bind %s\n", path);
+    }
+    addr.sun_family = AF_UNIX;
```
Please add `memset(&addr, 0, sizeof(struct sockaddr_un addr));` before using the structure.

ysmolsky commented on net.c:
```diff
@@ -32,118 +210,28 @@ make_server_socket(char *host, char *port)
+            errno = -r;
             return -1;
         }
         if (!r) {
```
Are sure that r==0 here in the case when systemd provided listen fd for the unix socket?

etanol replied:
[From `libsystemd`'s man pages][1]:

> ### Return value
>
> On failure, these calls return a negative errno-style error code. If the file descriptor is of the specified type and bound to the specified address, a positive return value is returned, otherwise zero.

[1]: https://www.freedesktop.org/software/systemd/man/sd_is_socket_inet.html#

ysmolsky replied:
Ah, so it is the same as with case for inet socket above. Sounds right then.

etanol commented on net.c:
```diff
@@ -7,19 +7,197 @@
+
+        if (verbose) {
+            char hbuf[NI_MAXHOST], pbuf[NI_MAXSERV], *h = host, *p = port;
+            r = getnameinfo(ai->ai_addr, ai->ai_addrlen,
```
Accident, my bad.

ysmolsky commented on net.c:
```diff
@@ -7,143 +7,237 @@
+            addrlen = sizeof(addr);
+            r = getsockname(fd, (struct sockaddr *) &addr, &addrlen);
+            if (!r) {
+                r = getnameinfo(ai->ai_addr, ai->ai_addrlen,
```
Original line is `             r = getnameinfo((struct sockaddr *) &addr, addrlen,`, your version looks too complicated. What is purpose of writing into `ai->ai_addr` instead of local var `addr`?

etanol replied:
Again, an accident.  I never had the intention to modify the original code for the listening socket setup.  I trusted the automatic rebase and Emacs three way merge more than I should have.

ysmolsky commented on net.c:
```diff
@@ -7,143 +7,237 @@
+    }
+    strncpy(addr.sun_path, path, maxlen);
+
+    r = bind(fd, (struct sockaddr *) &addr, sizeof(struct sockaddr_un));
```
I guess you meant `struct sockaddr_un` in both of casts.

etanol replied:
No, the cast is intentional (otherwise it wouldn't be necessary), see [bind(2)'s man page][1].

- Simple answer: it wouldn't compile otherwise, because warnings are treated as errors
- Longer answer: this is C-style poor man's polymorphism from the 80's, and also why the structure length needs to be provided too

[1]: https://manpage.me/index.cgi?q=bind&sektion=2

ysmolsky commented on net.c:
```diff
@@ -7,143 +7,237 @@
+    memset(&addr, 0, sizeof(struct sockaddr_un));
+    addr.sun_family = AF_UNIX;
+    if (strlen(path) > maxlen) {
+        warnx("truncating local socket path to: %.*s", (int)maxlen, path);
```
I propose that we `return -1` in case of truncated socket path. We should fail clearly ASAP of socket path is not correct. The test looked fishy, so I examined behaviour here, we should not allow this kind of error because when path was truncated then user definitely has an issue. We can keep the `strncpy` as safety measure though.

And test can check the returned results to check that long socket path or path to wrong directory won't be tolerated by the program.

etanol replied:
Totally agree.  Even with a warning, truncating the name may just lead to other unrelated and puzzling failures.

I didn't think that thoroughly before.

ysmolsky commented on testserv.c:
```diff
@@ -383,6 +430,41 @@ cttest_peek_not_found()
+    char name[90];
+    snprintf(name, sizeof(name), "/tmp/ctb-%d", getpid());
+    mustforksrv(name);
+    assert(exist(name));
```
These 5 lines above should be simplified. Please look at the ctdir() function. This functions returns path to the temp directory, so there is no need to do stuff with getpid(). I think that name for the socket path should be generated in the forking function.

etanol replied:
I initially wanted to to use that temporary directory, but I didn't search enough to figure out the `ctdir()` function was what I needed.

ysmolsky commented on testserv.c:
```diff
@@ -343,6 +376,20 @@ cttest_too_long_commandline()
+    const size_t maxlen = sizeof(addr.sun_path);
+    char real_name[maxlen];
+    snprintf(real_name, maxlen, "%s", NAME);
+    mustforksrv(NAME);
```
You could call `make_server_socket()` here...

ysmolsky commented on testserv.c:
```diff
@@ -141,30 +165,39 @@ kill_srvpid(void)
 // The child process serves until the SIGTERM is received by it.
 static int
-mustforksrv(void)
+mustforksrv(char *socket_file)
```
I would prefer not to mix UNIX socket into this function. This function returns an int, port value. Your change to this function looks confusing. I suggest to copy this function to mustforksrv_unix(). This function should return a socket path. And define macro SERVER_UNIX() similar to SERVER().
It is okay to have dup code in those functions for now. Simplicity is more important than elegance or cleverness.

etanol replied:
Sure.  But I'm blocked by #541.  So I won't be able to update this PR until #541 is merged.

ysmolsky replied:
Ok, let's wait for it. I won't plan any other merges too to not delay it further...

ysmolsky replied:
I have merged #541.

ysmolsky commented:
I really wish that Github notified me when there is a new push into the PR. Reviewing now.

ysmolsky commented on testserv.c:
```diff
@@ -166,16 +196,43 @@ mustforksrv(void)
     srvpid = fork();
-    if (srvpid < 0) {
-        twarn("fork");
+    if (srvpid > 0) {
```
Here the original code contained the check for srvpid < 0:
```
if (srvpid < 0) {
   twarn("fork");	
   exit(1);
}
```
The same check should happen in mustforksrv_unix

etanol replied:
Oops :-(

ysmolsky commented on testserv.c:
```diff
@@ -142,6 +166,12 @@ kill_srvpid(void)
+#define SERVER_UNIX() ({                        \
+            progname=__func__;                  \
+            char *path = mustforksrv_unix();    \
+            assert(exist(path));                \
```
Let's simplify this. We can move this assert into mustforksrv_unit() and then it can be made look similar to SERVER() macro.

ysmolsky commented on testserv.c:
```diff
@@ -166,16 +196,43 @@ mustforksrv(void)
-        printf("start server port=%d pid=%d\n", port, srvpid);
-        return port;
+        printf("start server socket=%s\n", path);
+        return path;
```
We could put assert after the line 234

ysmolsky commented on testserv.c:
```diff
@@ -363,6 +420,30 @@ cttest_peek_not_found()
+void
+cttest_peek_ok_unix()
+{
+    char *name = SERVER_UNIX();
```
Now it reads much better.

ysmolsky commented on testserv.c:
```diff
@@ -363,6 +420,30 @@ cttest_peek_not_found()
+{
+    // Twice, to trigger autoremoval
+    SERVER_UNIX();
+    SERVER_UNIX();
```
Do we expect that previous server instance should be still running here? If not we could kill it with kill_srvpid();

etanol replied:
Not really.  So yes, killing the server in between makes sense.

ysmolsky commented on doc/beanstalkd.1:
```diff
@@ -57,6 +57,9 @@ Show a brief help message and exit\.
 Listen on address \fIaddr\fR (default is 0\.0\.0\.0)\.
 .
 .IP
+When \fIaddr\fR starts with "unix:", the unprefixed value of will be used as the local filesystem path to create a UNIX socket instead of a TCP socket\. In this case the vale of \fB\-p\fR will be ignored\.
```
value of **it** will

ysmolsky replied:
vale -> value

ysmolsky commented on doc/beanstalkd.1.html:
```diff
@@ -121,6 +121,10 @@ <h2 id="OPTIONS">OPTIONS</h2>
 <dt class="flush"><code>-h</code></dt><dd><p>Show a brief help message and exit.</p></dd>
 <dt class="flush"><code>-l</code> <var>addr</var></dt><dd><p>Listen on address <var>addr</var> (default is 0.0.0.0).</p>
 
+<p>When <var>addr</var> starts with "unix:", the unprefixed value of will be
```
value of **it** will

ysmolsky commented on doc/beanstalkd.ronn:
```diff
@@ -59,6 +59,10 @@ and format of the `beanstalkd` protocol.
 * `-l` <addr>:
   Listen on address <addr> (default is 0.0.0.0).
 
+  When <addr> starts with "unix:", the unprefixed value of will be
```
value of **it** will

ysmolsky commented on doc/beanstalkd.ronn:
```diff
@@ -59,6 +59,10 @@ and format of the `beanstalkd` protocol.
 
+  When <addr> starts with "unix:", the unprefixed value of will be
+  used as the local filesystem path to create a UNIX socket instead of
+  a TCP socket. In this case the vale of `-p` will be ignored.
```
vale -> value

ysmolsky commented on doc/beanstalkd.1.html:
```diff
@@ -121,6 +121,10 @@ <h2 id="OPTIONS">OPTIONS</h2>
 
+<p>When <var>addr</var> starts with "unix:", the unprefixed value of will be
+used as the local filesystem path to create a UNIX socket instead of
+a TCP socket. In this case the vale of <code>-p</code> will be ignored.</p>
```
vale -> value

etanol commented:
No problem, glad to help and also get some closure.

ysmolsky commented:
Happy merged it. 

🎉

Fudoshiki commented:
hmm, how use that?

petri commented:
> hmm, how use that?

From the docs: 

`When addr starts with "unix:", the unprefixed value of it will be used as the local filesystem path to create a UNIX socket instead of a TCP socket.`

askurihin commented:
Can you please make a release, that will include this feature?

ysmolsky commented:
Yes, I am planning release very soon.
